examples/amm/src/lib.rs

```diff
@@ -0,0 +1,165 @@
+    type InitializationArgument = ();
+    type Parameters = Parameters;
+    type Operation = Operation;
+    type ApplicationCall = ApplicationCall;
+    type Message = Message;
+    type SessionCall = ();
+    type Response = ();
+    type SessionState = ();
+}
+
+impl ServiceAbi for AmmAbi {
+    type Query = Request;
+    type QueryResponse = Response;
+    type Parameters = Parameters;
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub enum OperationType {
+    Swap {
+        owner: AccountOwner,
+        input_token_idx: u32,
+        input_amount: Amount,
+        output_token_idx: u32,
+    },
+    AddLiquidity {
+        owner: AccountOwner,
+        token0_amount: Amount,
+        token1_amount: Amount,
+    },
+    RemoveLiquidity {
+        owner: AccountOwner,
+        input_token_idx: u32,
+        input_amount: Amount,
+        other_token_idx: u32,
+    },
+}
+
+scalar!(OperationType);
+
+/// Operations that can be sent to the application.
+#[derive(Debug, Serialize, Deserialize)]
+pub enum Operation {
+    ExecuteOperation { operation: OperationType },
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub enum Message {
+    ExecuteOperation { operation: OperationType },
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub enum ApplicationCall {
+    ExecuteOperation { operation: OperationType },
+}
+
+pub fn calculate_output_amount(
+    input_amount: Amount,
+    input_pool_balance: Amount,
+    output_pool_balance: Amount,
+) -> Result<Amount, AmmError> {
+    if input_pool_balance == Amount::ZERO || output_pool_balance == Amount::ZERO {
+        return Err(AmmError::InvalidPoolBalanceError);
+    }
+
+    let one_amount_bigint = BigUint::from_u128(<Amount as Into<u128>>::into(Amount::ONE))
+        .expect("Couldn't generate Amount::ONE in bigint");
+    let input_amount_bigint = BigUint::from_u128(<Amount as Into<u128>>::into(input_amount))
+        .expect("Couldn't generate input_amount in bigint");
+    let output_pool_balance_bigint =
+        BigUint::from_u128(<Amount as Into<u128>>::into(output_pool_balance))
+            .expect("Couldn't generate output_pool_balance in bigint");
+    let input_pool_balance_bigint =
+        BigUint::from_u128(<Amount as Into<u128>>::into(input_pool_balance))
+            .expect("Couldn't generate input_pool_balance in bigint");
+
+    // To always maintain the precision, whenever we multiply we divide by Amount::ONE
+    let numerator_bigint =
+        (input_amount_bigint.clone() * output_pool_balance_bigint) / one_amount_bigint.clone();
+    let denominator_bigint = input_pool_balance_bigint + input_amount_bigint;
+
+    if denominator_bigint.is_zero() {
+        return Err(AmmError::DivisionByZero);
+    }
+
+    // To always maintain the precision, whenever we divide, we always multiply
+    // by Amount::ONE before dividing
+    let output_amount_bigint = (numerator_bigint * one_amount_bigint) / denominator_bigint;
+    let output_amount = Amount::from(
```

Comment:
Keep in mind that `from` takes the number of attotokens. So `Amount::from(1)` is 10<sup>-18</sup>. But I think this is what we want here, we just shouldn't divide by `Amount::ONE` above.

Reply:
Maybe I'll change the `Amount::from(` instances to `Amount::attotokens(` to be clearer